core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -348,10 +347,15 @@ public synchronized ListenableFuture<QueryResults> waitForResults(long token, Ur
         return Futures.transform(futureStateChange, ignored -> getNextResult(token, uriInfo, targetResultSize), resultsProcessorExecutor);
     }
 
-    public synchronized void markResultsConsumedIfReady()
+    public void markResultsConsumedIfReady()
     {
-        if (!resultsConsumed && exchangeDataSource.isFinished()) {
-            queryManager.resultsConsumed(queryId);
+        if (!resultsConsumed) {
```

Comment:
Nit: it might be cleaner to return early if `resultsConsumed` is true than to nest the conditions and synchronizing blocks three levels deep.